cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
`charp` isn't guaranteed to be referencing memory under the control of arrow since `c_str() + s.length()` is one byte past the address of the last character, so we can't replace `charp == end` with `*charp == 0` generally speaking. Here's an example:

```c++
const char* str = "abcd";
size_t length = 4;
const char* start = str;
const char* end = str + length;
printf("%c", *end); // undefined behavior
```